hathor/indexes/rocksdb_tokens_index.py

```diff
@@ -0,0 +1,463 @@
+            if tx_output.can_melt_token():
+                # remove from melt index
+                self._remove_melt_utxo(token_uid, tx.hash, index)
+        else:
+            self._subtract_from_total(token_uid, tx_output.value)
+
+    def add_tx(self, tx: BaseTransaction) -> None:
+        # if it's a TokenCreationTransaction, update name and symbol
+        self.log.debug('add_tx', tx=tx.hash_hex, ver=tx.version)
+        if tx.version == TxVersion.TOKEN_CREATION_TRANSACTION:
+            from hathor.transaction.token_creation_tx import TokenCreationTransaction
+            tx = cast(TokenCreationTransaction, tx)
+            assert tx.hash is not None
+            self.log.debug('create_token_info', tx=tx.hash_hex, name=tx.token_name, symb=tx.token_symbol)
+            self._create_token_info(tx.hash, tx.token_name, tx.token_symbol)
+
+        if tx.is_transaction:
+            # Adding this tx to the transactions key list
+            assert isinstance(tx, Transaction)
+            for token_uid in tx.tokens:
+                assert tx.hash is not None
+                self._add_transaction(token_uid, tx.timestamp, tx.hash)
+
+        for tx_input in tx.inputs:
+            spent_tx = tx.get_spent_tx(tx_input)
+            self._remove_authority_utxo(spent_tx, tx_input.index)
+
+        for index in range(len(tx.outputs)):
+            self.log.debug('add authority utxo', tx=tx.hash_hex, index=index)
+            self._add_authority_utxo(tx, index)
+
+    def del_tx(self, tx: BaseTransaction) -> None:
+        for tx_input in tx.inputs:
+            spent_tx = tx.get_spent_tx(tx_input)
+            self._add_authority_utxo(spent_tx, tx_input.index)
+
+        for index in range(len(tx.outputs)):
+            self._remove_authority_utxo(tx, index)
+
+        if tx.is_transaction:
+            # Removing this tx from the transactions key list
+            assert isinstance(tx, Transaction)
+            for token_uid in tx.tokens:
+                assert tx.hash is not None
+                self._remove_transaction(token_uid, tx.timestamp, tx.hash)
+
+        # if it's a TokenCreationTransaction, remove it from index
+        if tx.version == TxVersion.TOKEN_CREATION_TRANSACTION:
+            assert tx.hash is not None
+            self._destroy_token(tx.hash)
+
+    def iter_all_tokens(self) -> Iterator[Tuple[bytes, TokenIndexInfo]]:
+        self.log.debug('seek to start')
+        it = self._db.iteritems(self._cf)
+        it.seek(bytes([_Tag.INFO.value]))
+        for (_cf, key), value in it:
+            key_any = self._from_key_any(key)
+            if key_any.tag is not _Tag.INFO:
+                break
+            self.log.debug('seek found', token=key_any.token_uid.hex())
+            info = self._from_value_info(value)
+            token_uid = self._from_internal_token_uid(key_any.token_uid)
+            token_index_info = RocksDBTokenIndexInfo(self, key_any.token_uid, info)
```

Comment:
`RocksDBTokenIndexInfo` class expects to have HTR `token_uid` as `\x00` or as the internal one? In this case here we are using the internal one, right? And in the method below it's not I guess.

Reply:
Actually `RocksDBTokenIndexInfo` will work with `\x00`, because it is converted later. But incidentally it will also work with the internal one. This was a rather late change I made, so I didn't stop to refactor it to make it clearer where we expect the internal one and where we expect the "public" one. It doesn't help that it's hard to use types to convey this (both are "bytes", but maybe it is possible to make a "newtype"  to make mypy distinguish both, although I hope this doesn't get too verbose).

Reply:
Fixed. It's not perfect because Python's `NewType` creates a subclass not an actually _new_ type, but it should be much clearer now, and I think there actually was a bug.

Reply:
I really like the `NewType` created and I feel the code is much easier to understand now.